### PR TITLE
RIA-1963: Joi Validation Extension

### PR DESCRIPTION
**Pull request by alexbaeza**
### Change description ###

RIA-1963: Extends Joi validation with a custom extension that check for mobile phones and formats them using google-libphonenumber.

Changed error message wording.
Added tests.

RIA-2280: Using a Unicode char as address for email should not be allowed. 
Tighten validation to prohibit Unicode emails.

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x ] No
```

### CODE REVIEW

File: app/utils/fields-validations.ts [hunk 1]
```diff
@@ -257,25 +261,27 @@ function statementOfTruthValidation(obj: object): null | ValidationErrors {
...
-    ['address-line-2']: Joi.string().optional().empty(''),
-    ['address-postcode']: Joi.string().optional().empty('').regex(postcodeRegex).messages({
+
+  const schema = joi.object({
+    ['address-line-1']: joi.string().required().messages({ 'string.empty': i18n.validationErrors.address.line1Required }),
+    ['address-town']: joi.string().required().messages({ 'string.empty': i18n.validationErrors.address.townCityRequired }),
+    ['address-county']: joi.string().optional().empty(''),
+    ['address-line-2']: joi.string().optional().empty(''),
+    ['address-postcode']: joi.string().regex(postcodeRegex).messages({
+      'string.empty': i18n.validationErrors.address.postcodeRequired,
```
Comment (chrisgrimble) [inline app/utils/fields-validations.ts]: This has been removed not think you have merged in master changes incorrectly

File: extensions/joi/mobile-number.ts [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+ * string.mobilePhoneNumber.invalid.string : if string supplied was invalid
+ * string.mobilePhoneNumber.invalid.mobile: if the string supplied was not a mobile phone
+ */
+module.exports = joi => {
+  return ({
+    base: joi.string(),
+    type: 'mobilePhoneNumber',
+    messages: {
+      'string.mobilePhoneNumber.invalid.string': '"{{#label}}" did not seem to be a valid phone number',
+      'string.mobilePhoneNumber.invalid.mobile': '"{{#label}}" did not seem to be a valid mobile phone number'
```
Comment (chrisgrimble) [inline extensions/joi/mobile-number.ts]: Needs to be in the internationalisation cannot have content here

File: extensions/joi/mobile-number.ts [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+          throw new Error('The phone supplied did not seem to be a valid mobile phone number');
+        }
+
+        return { value: phoneUtil.format(mobilePhoneNumber, format) };
+
+      } catch (err) {
+        const knownErrors = [
+          {
+            error: 'string.mobilePhoneNumber.invalid.string',
+            errorMessages: [ 'The string supplied did not seem to be a phone number', 'Phone number too short after IDD' ]
```
Comment (chrisgrimble) [inline extensions/joi/mobile-number.ts]: Not too sure how Joi works but again do these need ti be internationalised

File: extensions/joi/mobile-number.ts [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+          },
+          {
+            error: 'string.mobilePhoneNumber.invalid.mobile',
+            errorMessages: [ 'The phone supplied did not seem to be a valid mobile phone number' ]
```
Comment (chrisgrimble) [inline extensions/joi/mobile-number.ts]: Same as above

File: app/utils/fields-validations.ts (hunk 1 above)
Comment (alexbaeza) [reply app/utils/fields-validations.ts]: I'll double-check, Thank you Chris

File: extensions/joi/mobile-number.ts (hunk 3 above)
Comment (alexbaeza) [reply extensions/joi/mobile-number.ts]: Let's discuss this offline, I'll explain.

File: app/utils/fields-validations.ts [hunk 5]
```diff
@@ -28,8 +32,8 @@ function validate(obj: object, schema: any, abortEarly: boolean = false): Valida
 }
 
 function textAreaValidation(text: string, theKey: string): ValidationErrors | null {
-  const schema = Joi.object({
-    [theKey]: Joi.string()
+  const schema = joi.object({
```
Comment (t0nin0s) [inline app/utils/fields-validations.ts]: I would probably only use the extended version of joi for those cases needed...so just for phone number

File: app/utils/fields-validations.ts [hunk 6]
```diff
@@ -118,63 +122,63 @@ function dateValidation(obj: any, errors): boolean | ValidationErrors {
...
-      is: Joi.alternatives()
+    'text-message-value': joi.when('selections', {
+      is: joi.alternatives()
         .try(
-          Joi.array().items(Joi.string().valid('email'), Joi.string().valid('text-message')),
-          Joi.string().valid('text-message')).required(),
-      then: Joi.string()
+          joi.array().items(joi.string().valid('email'), joi.string().valid('text-message')),
+          joi.string().valid('text-message')).required(),
+      then: joi.mobilePhoneNumber().format('e164')
```
Comment (t0nin0s) [inline app/utils/fields-validations.ts]: so I would probably do...
Joi.extend(MobilePhoneNumberExtension).format.....

Comment (alexbaeza) [reply app/utils/fields-validations.ts]: Thank Toni, I will make the suggested change on the next commit

**Comment by alexbaeza** [comment]
Closing this PR and moved to: https://github.com/hmcts/ia-aip-frontend/pull/112
Due to an issue in jenkins.

**Pull request closed without merging**